### PR TITLE
Work on whistle algorithm

### LINKED ISSUE #470: The whistle algorithm can produce events outside of the bandwidth bounds
## Actual behaviour:

When using the `OnebinTrackAlgorithm` with `MinHertz` and `MaxHertz` set, events can be produced outside of the bandwidth limit.

## Expected behavior:

Events should not be produced outside of the bandwidth limits

## How to reproduce this bug:

1. Define a generic recogniser to have a whistle detection with some limits
2. Ensure you're using a high window frame (like 2048)
3. In some cases you'll note events that are output below the lower limit.

## Additional Details

image

In the case above events were reported out side the bandwidth limits.

- `MinHertz` was set to `2850`
- Events were produced with bounds of  `"LowFrequencyHertz": 2770.0, "HighFrequencyHertz": 2780.0,`

### PR DESCRIPTION
**Pull request by towsey**
Issue #470 Endeavour to fix error where event was found outside search frequency band. I believe this error was caused by the rounding of the search bounds from Hertz to a frequency bin. I have endeavored to tighten up the conversion of Hertz bound to frequency bin. Also added amore components to the unit test for whistle events.
Also added two lines to write spectrogram results for visualization. This is a temporary fix - to be removed later.

# Incorporate repair to whistle algorithm

## Changes
Main change is the how the search bounds in Hertz are converted to bin search bounds.
Also added more components to the unit test.

## Issues
It is possible that this "error" could occur again in situation where the lower search bound is very close to an actual whistle event. I did duplicate this error with the original code but have not been able to duplicate it with the revised code - so hopefully the error is fixed. If the error occurs again I believe it is due to the frequency conversion as described above and is not a serious error.
 
## Visual Changes
Two lines were added to the unit test in order to produce a debug image. These can be deleted when the debug spectrogram gets written to locatable folder.

Fixes #470

### CODE REVIEW

**Comment by towsey** [comment]
I have fixed two broken unit tests. The whistle test was broken due to introduction of two new variables to remove an ambiguous dual use of the variables MinHertz and MaxHertz.
Request @Truskinger to check what has been done before proceeding to remove this ambiguity in other generic syllable types.

**Comment by atruskie** [comment]
@towsey 

I haven't reviewed this properly yet, but some notes:

- What is the difference between `[Min|Max]Hertz` and `Searchband[Min|Max]Hertz`?
    - The documentation for these new properties is not aiding my understanding either...
    ```
    Gets or sets the bottom bound of the rectangle. Units are Hertz.
    ```
    vs
    ```
    Gets or sets the minimum bandwidth, units = Hertz.
    ```
- The validation in `CommonParameters` probably should not be disabled: https://github.com/QutEcoacoustics/audio-analysis/pull/477/files#diff-090545b4a4071d03cbd9d28356981e8bb5b9a9a1a64a1160b57070f771ead268R98-R99
  Other algorithms assume these values are not `null` so your change would affect more than just the whistle algoirthm.
- Do note I'll me integrating my interval library into AP soon. This will remove all `min|max` properties which should allow for more natural representations of parameters. e.g.
    ```yaml
    SearchBandHertz: "[400, 2000]"
    ScoreThrehsold: ">0.6"
    ```


My understanding is the existing `[Min|Max]Hertz` values are treated as search limits. They're sometimes also used to set the size of resulting event afterwards (if no more pertinent information is available). Thus, your new parameters are inconsistent with the other algorithms?

If I've misunderstood, then we should set up a call?

**Comment by towsey** [comment]
Sorry - I am having a bit of trouble getting internet connection here so I upload even when not finished. I rewritten the summaries for the two new variables and hopefully they explain what is intended. Min/MaxHertz are event parameters where as SearchbandMin/MaxHertz are algorithm parameters. I am not sure that I have placed them in the best position. 

The new interval library looks good.

I commented out the two validation lines to get the new variables working. I am not sure where the most logical place is to put the new variables. They will be common to most algorithms but not all. The Blob algorithm and the Harmonics algorithm would not currently need the distinction. However I am thinking to write a new Blob2 event recognizer which would detect blob events of limited bandwidth anywhere in a wide search band. This would be a better way to do your Bell Bird recognizer.

If this does not make the issues clearer then we will have to have a call. But it is not easy here on French Island. We are leaving FI on Thursday and driving to Albury on Friday.

**Comment by atruskie** [comment]
Ok the new documentation makes sense.

To summarize: the new definition for `[Min|Max]Hertz` set limits on event bandwidth size, while `Searchband[Min|Max]Hertz` set limits on where the algorithm can search.

Follow up questions:

1. Can't post-processing do the job of the new  `[Min|Max]Hertz`?
2. Previously `[Min|Max]Hertz` operated the same as `Searchband[Min|Max]Hertz`, why change the name?
3. With your current changeset you've only changed two of the algorithms' names?
4. Even though you've provided a distinction between the two parameters (`[Min|Max]Hertz` and `Searchband[Min|Max]Hertz`) they both aren't used at the same time, suggesting the distinction is not needed?

For reference here are all the places where `[Min|Max]Hertz` are currently used as a bandwidth limit to the search:

- https://github.com/QutEcoacoustics/audio-analysis/blob/16580beb3d038dff5837dec9a8c7478bf417be6a/src/AudioAnalysisTools/Tracks/ForwardTrackAlgorithm.cs#L87
- https://github.com/QutEcoacoustics/audio-analysis/blob/16580beb3d038dff5837dec9a8c7478bf417be6a/src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs#L84
- https://github.com/QutEcoacoustics/audio-analysis/blob/d743f1115063f24ecdc53dab3826fd48ab46217c/src/AudioAnalysisTools/Tracks/OneframeTrackAlgorithm.cs#L87
- https://github.com/QutEcoacoustics/audio-analysis/blob/d743f1115063f24ecdc53dab3826fd48ab46217c/src/AudioAnalysisTools/Tracks/UpwardTrackAlgorithm.cs#L93
- Blob event is weirder but still the same: https://github.com/QutEcoacoustics/audio-analysis/blob/e1a6784f4104aad7aa44d3b736d442295b584a99/src/AudioAnalysisTools/DSP/SNR.cs#L508
- https://github.com/QutEcoacoustics/audio-analysis/blob/e1a6784f4104aad7aa44d3b736d442295b584a99/src/AudioAnalysisTools/Harmonics/HarmonicAnalysis.cs#L43-L56
- https://github.com/QutEcoacoustics/audio-analysis/blob/e1a6784f4104aad7aa44d3b736d442295b584a99/src/AudioAnalysisTools/Ocillations/Oscillations2012.cs#L199
    
All these uses were consistent. Thus, this change is a partially implemented suggestion to change all uses?


Can I suggest that no change to the names are needed currently?

Or as another suggestion, leave  `[Min|Max]Hertz` as is, and introduce a new more specific parameter for whistle algorithm (all though again, it does not look like it is needed).

**Comment by atruskie** [comment]
We're communicating offline with bad internet connections so I'm including a summary of our convsation here for prosperity:

> Concerning the issue of the generic recognizer Hertz parameters, in the most general case a profile needs at least four Hertz parameters, two to set the search bounds and two to set the event bandwidth bounds. After an event is found, its actual location in frequency domain is then assigned to the MinHertz and Max Hertz parameters.
> 
>  
> 
> Historically with the blob recognizer, the search bounds and event bounds were conflated. However conflating these parameters is very limiting as you discovered with the Bellbird recognizer. When you explained what you were doing with the Bellbird recognizer, I immediately understood that we need a new algorithm with the four parameters as described above. I was intending to call it Blob2 unless you can think of a more sensible name.
> 
>  
> 
> I had already implemented something like this with the whistle recognizer, the only difference being that the event bandwidth of a whistle is by default three frequency bins wide. So no need to set this parameter. But in implmenting the whistle, I reused the Min/MaxHertz parameters even though their actual meaning was now something different.
> 
>  
> 
> In one of your issues I saw that you talked about ambiguity and I thought this was the problem you were referring to. Now realise it is something different. But I decided to fix up the names of the whistle and whip parameters to make them consistent with what I stated above. The click and chirp algorithms have yet to be done. I stopped because I am not sure how best to make these changes with the current class hierarchy. That is why also commented out the two verification lines - just to get things working.
> 
>  
> 
> I am pretty sure that the oscillations reconizer already searches a frequency band but I need to check that again.
> 
>  
> 
> With the harmonics recognizer, the searchband will also be the event bounds because the implementation of the DCT algorithm offers more flexible. In other words, the profile only needs the two searchband bounds because the DCT algorithm can (within limits) detect a set of harmonics anywhere in the searchband.
> 
>  
> 
> And with regard to your final point - can we not use the post-processing to deal with bandwidth? The short answer in my opinion is no. The profiles define the parameters required to find syllable/primitive acoustic events. THe post-processing is all about combining syllable/primitives to recognize multi-syllable calls. After the syllables have been combined we need the time/frequency filters to remove combined events that fall outside the call constraints. These are different from the syllable constraints.
> 
>  
> 
> I need your advice as how to incorporate these changes into the current class hierarchy or how to change the hierarchy. But for clarity, I believe we need to make all the above distinctions.
> 
>  
> 
> Cheers, Michael

---

My response:

Like I said in the issue I think the simplest thing is not change the meaning of [Min|Max]Hertz simply because it's usage is currently consistent and widespread. Let's treat [Min|Max]Hertz  as search bounds - it's just the simplest way forward, even if their current meaning was not your original intended meaning.
 

> two to set the event bandwidth bounds.

As you pointed out, most algorithms at the moment simply default to using the search range as the bounds of the event. When more pertinent information is available (e.g. whistle algorithm defaults to three bins) then more precise bounds are reported.
 

I think rather than changing [Min|Max]Hertz to mean event bandwidth limit we should keep the definition as search limits and for algorithms that need event bandwidth limits we should add a new parameter:  MinEventBandwidthHertz and MaxEventBandwidthHertz (or with an interval, EventSizeHertz: "[100, 300]")

 

I agree that [Min|Max]Hertz probably is the best term to describe the event bandwidth limit, however, that ship has sailed.
 


> one of your issues I saw that you talked about ambiguity

Can you link me to it? I can add more context then. 
There's https://github.com/QutEcoacoustics/audio-analysis/issues/471 but that's for harmonics. And https://github.com/QutEcoacoustics/audio-analysis/issues/461 but that's just a cleanup of unused things.
 

> I was intending to call it Blob2 unless you can think of a more sensible name.

Yes I absolutely want to make new algorithms. We should come up with a more descriptive name though. Blob2 can work as a beta/dev name until the semantics and differences are clear.

To summarize, you did:
 

	Changed [Min|Max]Hertz to Searchband[Min|Max]Hertz
	Defined the new [Min|Max]Hertz to be an event size limit
	The problem being the usage of [Min|Max]Hertzis now inconsistent

 I'm suggesting:
 

	Do not change [Min|Max]Hertz  
	For algorithms that need event size limits add a new parameter named [Min|Max]EventSizeHertz

Because that change is far smaller

**Comment by towsey** [comment]
I have removed the names of the two search bound variables I created and in addition added comments to the existing Min/MaxHertz variables indicating that they are actually functioning as search bounds.

File: src/AudioAnalysisTools/CommonParameters.cs [hunk 1]
```diff
@@ -95,8 +101,8 @@ public abstract class CommonParameters : IValidatableObject
 
         public virtual IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
         {
-            yield return this.MinHertz.ValidateNotNull(nameof(this.MinHertz));
-            yield return this.MaxHertz.ValidateNotNull(nameof(this.MaxHertz));
+            //yield return this.MinHertz.ValidateNotNull(nameof(this.MinHertz));
+            //yield return this.MaxHertz.ValidateNotNull(nameof(this.MaxHertz));
```
Comment (atruskie) [inline src/AudioAnalysisTools/CommonParameters.cs:105]: Is there a reason this validation is disabled?

Revert change or delete validation (and explain).

File: src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs [hunk 2]
```diff
@@ -66,8 +71,22 @@ public static class OnebinTrackAlgorithm
...
+            if (minSearchBin < 1)
+            {
+                minSearchBin = 1;
+            }
+
+            // set top search bin allowing for the top sideband.
+            int maxSearchBin = (int)Math.Floor(parameters.SearchbandMaxHertz.Value / binWidth) - 1;
+            if (maxSearchBin > binCount - 6)
+            {
+                maxSearchBin = binCount - 6;
```
Comment (atruskie) [inline src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs:86]: What is this magic number `6`? Is that the "sideband"?  Extract to a constant?

File: src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs [hunk 3]
```diff
@@ -79,20 +98,23 @@ public static class OnebinTrackAlgorithm
...
                     }
 
-                    // here we define the amplitude profile of a whistle. The buffer zone around whistle is five bins wide.
+                    // here we define the amplitude profile of a whistle.
+                    // The buffer zone around centre of whistle is five bins wide. Ignore bins -2 and +2
                     var bandIntensity = ((sonogramData[tf, bin - 1] * 0.5) + sonogramData[tf, bin] + (sonogramData[tf, bin + 1] * 0.5)) / 2.0;
                     var topSidebandIntensity = (sonogramData[tf, bin + 3] + sonogramData[tf, bin + 4] + sonogramData[tf, bin + 5]) / 3.0;
                     var netAmplitude = 0.0;
-                    if (bin < 4)
+                    if (bin < 5)
```
Comment (atruskie) [inline src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs:117]: More magic numbers

File: src/AudioAnalysisTools/Tracks/OnebinTrackParameters.cs [hunk 4]
```diff
@@ -20,6 +20,16 @@ namespace AnalysisPrograms.Recognizers.Base
...
+        /// <summary>
+        /// Gets or sets a value indicating the minimum Hertz value of the search band.
+        /// </summary>
+        public int? SearchbandMinHertz { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating the maximum Hertz value of the search band.
+        /// </summary>
+        public int? SearchbandMaxHertz { get; set; }
+
```
Comment (atruskie) [inline src/AudioAnalysisTools/Tracks/OnebinTrackParameters.cs:32]: These do not appear to be used. They're redundant? Remove?

File: src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs [hunk 5]
```diff
@@ -66,8 +71,22 @@ public static class OnebinTrackAlgorithm
...
+
+            // set lower frequency bins of the search band
+            int minSearchBin = (int)Math.Floor(parameters.SearchbandMinHertz.Value / binWidth);
+            if (minSearchBin < 1)
+            {
+                minSearchBin = 1;
+            }
+
+            // set top search bin allowing for the top sideband.
+            int maxSearchBin = (int)Math.Floor(parameters.SearchbandMaxHertz.Value / binWidth) - 1;
```
Comment (atruskie) [inline src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs:83]: The incorrect parameters `SearchbandMinHertz` and `SearchbandMaxHertz` are still being used here

**Review by atruskie — CHANGES_REQUESTED** [review]

File: src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs (hunk 5 above)
Comment (towsey) [reply src/AudioAnalysisTools/Tracks/OnebinTrackAlgorithm.cs]: I have made the corrections as requested. I have also made an extensive class summary for the OneBinTrackAlgorithm class i.e. the class used to find whistles. I hope this will be helpful in understanding the algorithm and explaining the "magic numbers". In fact all the track algorithms  require that a lot of thought be given to the choice of sample rate, frame size and frame step. The typical values are not always appropriate depending on the target call.

**Review by atruskie — APPROVED** [review]
